enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,108 @@
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM malfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect inactive cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
+to: `Operator checking for stale status, the active operator will reset this message: <original message>`.
+
+After the message is changed, as it suggests, the active operators should reset this message back to the original
+message, removing the "checking" part. This shows a healthy, active operator.
+
+Suppose the operator is not active and does not reset the message back to the original after 10 minutes. In that case,
+the controller will change the message again
+to: `Operator has not updated this condition for more than 20 minutes, last known condition state was "<False/True>", original message: <original message>`
+. In addition to changing the message, this time, the controller will also change the condition status from False or
+True to **Unknown**.
```

Comment:
@deads2k your slack message says we'd set the operator to "A=unknown, D=true, P=unknown".  Assuming that's the plan, can we get the EP updated to reflect that behavioral mapping?  (And also what it would do for the upgrading condition).

Also i would posit that this new logic should *not* set Available=unknown if Available is currently set to False.  Though i can see arguments for why unknown might be more appropriate, reducing the severity of the status(I would argue unavailable is more severe than unknown, in terms of how admins are likely to treat it/have alerts configured) doesn't seem appropriate to me.  And if the operator isn't updating its status, it seems pretty unlikely that it is actually Available.